src/piton/labelers/core.py

```diff
@@ -0,0 +1,381 @@
+from __future__ import annotations
+
+import pprint
+import collections
+import datetime
+import json
+import pickle
+import os
+from collections.abc import MutableMapping
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import (
+    Any,
+    DefaultDict,
+    Dict,
+    List,
+    Literal,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from .. import Event, Patient
+from ..extension import datasets as extension_datasets
+
+import numpy as np
+
+@dataclass
+class TimeHorizon:
+    start: datetime.timedelta
+    end: datetime.timedelta
+
+@dataclass
+class SurvivalValue:
+    event_time: int # TODO - rename
+    is_censored: bool # TRUE if this patient was censored
+
+LabelType = Union[
+    Literal["boolean"],
+    Literal["numeric"],
+    Literal["survival"],
+    Literal["categorical"],
+]
+
+VALID_LABEL_TYPES = ["boolean", "numeric", "survival", "categorical"]
+
+
+# Make sure this is pickleable
+class Label:
+    """
+        An individual label for a particular patient at a particular time.
+    """
+
+    __slots__ = [
+        "time", # Arbitrary timestamp (datetime.datetime)
+        "label_type",
+        "value",
+    ]
+
+    label_type: LabelType
+
+    def __init__(
+        self,
+        time: datetime.datetime,
+        value: Union[bool, int, float, SurvivalValue],
+        label_type: LabelType,
+    ):
+        """Construct a label for datetime `time` and value `value`. 
+
+        Args:
+            time (datetime.datetime): Time in this patient's timeline that corresponds to this label
+            value (Union[bool, int, float, SurvivalValue]): Value of label. Defaults to None.
+            label_type (LabelType): Type of label. Must be an element in `VALID_LABEL_TYPES`.
+        """
+        assert label_type in VALID_LABEL_TYPES
+        self.time = time
+        self.label_type = label_type
+        self.value = value
```

Comment:
We should assert that self.value matches what it should be from label_type

if label_type == "boolean":
   assert self.value in (True, False)